modules/build/src/main/scala/scala/build/ResourceMapper.scala

```diff
@@ -0,0 +1,122 @@
+package scala.build
+
+import scala.collection.Map
+import scala.jdk.CollectionConverters._
+import scala.build.internal.Constants
+
+trait BaseResourceMapper {
+  private def readMappingIfExists(mappingFile: os.Path): Option[Map[os.Path, os.Path]] =
+    if (os.exists(mappingFile)) {
+      val mappingContent = os.read(mappingFile)
+      if (mappingContent.isEmpty()) Some(Map.empty)
+      else Some(
+        mappingContent.split('\n').map { pair =>
+          val sep = pair.split(',')
+          (os.Path(sep(0)), os.Path(sep(1)))
+        }.toMap
+      )
+    }
+    else None
+
+  private def writeMapping(mappingPath: os.Path, mapping: Map[os.Path, os.Path]) = {
+    val mappingContent = mapping.map { case (inputPath, outputPath) =>
+      s"$inputPath,$outputPath"
+    }.mkString("\n")
+    os.write.over(mappingPath, mappingContent)
+  }
+
+  protected def copyResourcesToDirWithMapping(
+    build: Build.Successful,
+    mappingFilePath: os.Path,
+    mappingFunction: Build.Successful => Map[os.Path, os.Path]
+  ): Unit = {
+
+    def isInDirectory(output: os.Path, filePath: os.Path) = {
+      val outputFullPath = output.toNIO.iterator().asScala.toSeq
+      val fileFullPath   = output.toNIO.iterator().asScala.toSeq
+      fileFullPath.startsWith(outputFullPath)
+    }
```

Comment:
I believe this should work (and this can probably be inlined then):
```suggestion
    def isInDirectory(output: os.Path, filePath: os.Path) =
      filePath.startsWith(output)
```